### PR TITLE
Period Era Re-Enactment: Autopunctuation 2

### PR DESCRIPTION
**Pull request by Ephemeralis**
## About The Pull Request

This PR introduces basic automatic punctuation for the following circumstances on living mobs only, and **only for says** (/me or emotes that aren't special say-emotes are unaffected):

- If a chat message does not end in any kind of punctuation (like a period (.), a comma (,), a question mark (?), an exclamation mark (!) or a hyphen (-)), a period is automatically added to the end.
- Any kind of lowercase 'i' between word boundaries is automatically converted to uppercase.

**To be completely clear: this does not affect PDA's, asterisk emotes (\*emotes), full emotes, announcements or basically any other non-speaking form of communication.** It affects speech, so says and radio chatter *only*. PDAspeak enjoyers can continue to sling slang at one another without The Man moderating their grammar.

I've strategically chosen these two changes to produce what I believe to be most comfortable compromise between an easy chatting experience and having dialogue look clean, while also allowing for people to continue to flavour their character's speech patterns as appropriate.

There is *one* caveat: if you make use of in-game bold/italics/underline formatting, the autopunct function this PR adds will **not** add an additional end-of-line period if your message does not contain one. For example: `|I can't believe you've done this|` will not have a period added - you'll need to do this yourself.

I've confirmed with our unofficial TGcode liaison (goof) that regex filters have negligible performance impacts (and that any concerns that they did was largely what he describes as 'cargo cult' behavior). See for yourself on /tg/'s Discord [here](https://discord.com/channels/326822144233439242/326831214667235328/1186135072617799690), if you like.

tl;dr: easier and faster Good Chatting (tm) with likely little performance impact

## How This Contributes To The Skyrat Roleplay Experience

Dialogue becomes much cleaner and coincidentally, easier to type. Reach across your keyboards for the period key no more - chat as you would on Discord, and have the game prettify most of it for you. Never again suffer the tyranny of reaching for your shift-key mid conversation and adding four seconds onto your message typing time.

It is surprisingly comfortable. I encourage doubters to clone this locally and see for themselves.

Importantly, it makes interacting *easier*. If you're a slow typer trying to meet standards, this can legitimately shave off seconds per message - if you're sending hundreds of messages a round, you're responding more quickly and so are the people participating with you. It's a genuinely objective improvement in ways beyond just 'chat look good'.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

https://github.com/Skyrat-SS13/Skyrat-tg/assets/966289/123f5f06-b76a-4e42-af8d-ebda8619e7b7

</details>

## Changelog

:cl: yooriss
add: Talking now automatically capitalizes any lowercase 'i' in the middle of sentences, and ends sentences in a period if no other punctuation is present.
/:cl:

### CODE REVIEW

**Comment by R9b1t** [comment]
i personally dont see this as good as it removes the ability for someone to control their own speech. i realize that not having a period or not capitalizing an 'i' is an ick for some people

but why force it when someone might be intentional and not want the chat system to force them along. at the very least maybe add a toggle option so people can CHOOSE if they want auto-punc?

**Comment by nikothedude** [comment]
ive always argued against this

with autopunc you lose the ability to distinguish people who are trying to the people who arent, and as ive personally experienced on like, sojourn, it just makes everyone around you, even the people who ARE trying, seem just as bad as people who arent trying

File: modular_skyrat/master_files/code/_globalvars/regexes.dm [hunk 1]
```diff
@@ -0,0 +1,5 @@
+//Any EOL char that isn't appropriate punctuation
```
Comment (nikothedude) [inline modular_skyrat/master_files/code/_globalvars/regexes.dm:1]: convert to autodoc, autodoc functions on global things iirc

**Comment by Nerev4r** [comment]
if everyone around them sees 'nobody punctuating and everyone writing as sloppily as possible' they will not be any encouraged to step their game up. but if everyone around them (thanks to automated yooriss code) is typing correctly, with punctuation and everything, they will be more inclined to step their game up. every other roleplay server has this and it's also true there.
you do not need to 'filter' anyone and you do not need to 'distinguish from anyone' and it does not 'make you lose control of what you're typing.' it steps the standard up and to quote the author, in effect all this PR does is make it easier to type quickly and cleanly in high-action situations.

**Comment by nikothedude** [comment]
> if everyone around them sees 'nobody punctuating and everyone writing as sloppily as possible' they will not be any encouraged to step their game up. but if everyone around them (thanks to automated yooriss code) is typing correctly, with punctuation and everything, they will be more inclined to step their game up. every other roleplay server has this and it's also true there. 

i countered this argument on dsicord btw: this is false, because the viewer intrinsically knows they dont HAEV to actually top the seen message, because they know they can talk like I am now, and will automatically meet their writing quality

**Comment by Nerev4r** [comment]
> > if everyone around them sees 'nobody punctuating and everyone writing as sloppily as possible' they will not be any encouraged to step their game up. but if everyone around them (thanks to automated yooriss code) is typing correctly, with punctuation and everything, they will be more inclined to step their game up. every other roleplay server has this and it's also true there.
> 
> i countered this argument on dsicord btw: this is false, because the viewer intrinsically knows they dont HAEV to actually top the seen message, because they know they can talk like I am now, and will automatically meet their writing quality

cap

**Comment by chemistrymain2** [comment]
> > if everyone around them sees 'nobody punctuating and everyone writing as sloppily as possible' they will not be any

It visually makes the chat nicer and there's no reason not to have it when several other servers do, all of them HRP.

**Comment by kittysmooch** [comment]
does this affect PDA texts at all? i feel like we should leave the allowance for those HRPers who go their textspeech vibed out.

**Comment by Ephemeralis** [comment]
> does this affect PDA texts at all? i feel like we should leave the allowance for those HRPers who go their textspeech vibed out.

it touches nothing but says - PDA use an entirely different ingress afaik

**Comment by Iamgoofball** [comment]
hello, you appear to have forgotten to remove the punctuation at mob/living/Hear() for the purposes of runechat; runechat looks bad with "." as ending punctuation, please ensure you're intercepting the message and removing the added "." at that point

**Comment by nikothedude** [comment]
> hello, you appear to have forgotten to remove the punctuation at mob/living/Hear() for the purposes of runechat; runechat looks bad with "." as ending punctuation, please ensure you're intercepting the message and removing the added "." at that point

wtf are you talking about? i have ended my sentences thousands of times with periods and theres literally nothing wrong with it

**Comment by ghost** [comment]
> If this impacts your character's speech characterization in any way, please let us know in #development-code on Discord.

I'd remove this line in the CL, issues are meant to be reported on github, not in discord.

**Comment by Ephemeralis** [comment]
> > If this impacts your character's speech characterization in any way, please let us know in #development-code on Discord.
> 
> I'd remove this line in the CL, issues are meant to be reported on github, not in discord.

Gone ahead and scrubbed this out.

**Comment by HoodMilk** [comment]
Removes the ability to add delays into your speech, kinda lame. Sometimes I'll send the first half of a message and then begin typing the second, all the same sentence, for dramatic effect and also so people don't have to watch me type for 20 seconds.

**Comment by Nerev4r** [comment]
> Removes the ability to add delays into your speech, kinda lame. Sometimes I'll send the first half of a message and then begin typing the second, all the same sentence, for dramatic effect and also so people don't have to watch me type for 20 seconds.

just put a hyphen at the end vro

**Comment by Ephemeralis** [comment]
> Removes the ability to add delays into your speech, kinda lame. Sometimes I'll send the first half of a message and then begin typing the second, all the same sentence, for dramatic effect and also so people don't have to watch me type for 20 seconds.

You can hyphenate (and probably should, for readability's sake) partial responses. This does not stop you from doing that.

**Comment by chemistrymain2** [comment]
The lengths people are going to to justify ending their sentences like this

Instead of like this.

You people are insane jfc every single roleplay server has this added by default.

**Review by tf-4 — COMMENTED** [review]
Should be a pref toggle
